### PR TITLE
New LDAPProvider for user_ldap

### PR DESCRIPTION
**Pull request by GitHubUser4234**
Please refer to owncloud/core#23992

### CODE REVIEW

**Comment by MorrisJobke** [comment]
best would be to wait until this is integrated into owncloud, because we regularly fetch the changes from there. Otherwise we will fix stuff here differently from what is fixed in ownCloud and merging this then in the end will be a nightmare.

@GitHubUser4234 Is this okay for you? I would then close this PR.

**Comment by GitHubUser4234** [comment]
@MorrisJobke: We switched to NC and it is high priority for us to have it included in NC 10. Waiting would put this target at risk. For OC we won't push for a merge there in the near future.

**Comment by MorrisJobke** [comment]
> @MorrisJobke: We switched to NC and it is high priority for us to have it included in NC 10.

We already released a beta release of NC 10 and will not include new features anymore in the stable10 branch. This needs to wait for NC 11

**Comment by MorrisJobke** [comment]
cc @blizzz

**Comment by blizzz** [comment]
> > @MorrisJobke: We switched to NC and it is high priority for us to have it included in NC 10.

> We already released a beta release of NC 10 and will not include new features anymore in the stable10 branch. This needs to wait for NC 11 

I am not aware we announced a feature freeze, did we? Also, because we still lack other planned features for 10.

**Comment by andreas-p** [comment]
AFAICS there's no setPassword functionality implemented yet?
I've written a somewhat hackyish app allowing the user to change his own password (userPassword with several crypt options including SSHA, as well as sambaNTPassword in case it's a sambaSamAccount) which I could contribute.

**Comment by GitHubUser4234** [comment]
@andreas-p : good news, a setPassword functionality is in the works: #520  
Password hashing won't be in though because 

1. it prevents password policy enforcement
2. it can be done at (LDAP) server side

**Comment by andreas-p** [comment]
Well, if sambaNTpassword isn't set correctly in parallel it's unusable for me....
AFAIK there's no ldap_passwd or similar function, that's why I implemented hashing myself, necessary for Samba anyway.

**Comment by blizzz** [comment]
Integration Tests succeeding :white_check_mark:

File: apps/user_ldap/tests/Jobs/CleanUpTest.php [hunk 1]
```diff
@@ -37,7 +37,7 @@ public function getMocks() {
 				->getMock();
 		$mocks['ocConfig']    = $this->getMock('\OCP\IConfig');
 		$mocks['db']          = $this->getMock('\OCP\IDBConnection');
-		$mocks['helper']      = $this->getMock('\OCA\User_LDAP\Helper');
+		$mocks['helper']      = new \OCA\User_LDAP\Helper();
```
Comment (blizzz) [inline apps/user_ldap/tests/Jobs/CleanUpTest.php]: this change makes the unit test break

File: apps/user_ldap/appinfo/update.php [hunk 2]
```diff
@@ -0,0 +1,22 @@
+<?php
+/**
+ * @author Roger Szabo <roger.szabo@web.de>
+ *
+ * @copyright Copyright (c) 2016, ownCloud, Inc.
```
Comment (blizzz) [inline apps/user_ldap/appinfo/update.php]: That line should go

File: apps/user_ldap/lib/IUserLDAP.php [hunk 3]
```diff
@@ -0,0 +1,48 @@
+<?php
+/**
+ * @author Roger Szabo <roger.szabo@web.de>
+ *
+ * @copyright Copyright (c) 2016, ownCloud, Inc.
```
Comment (blizzz) [inline apps/user_ldap/lib/IUserLDAP.php]: This line should go, too

File: apps/user_ldap/lib/LDAPProviderFactory.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
+<?php
+/**
+ * @author Roger Szabo <roger.szabo@web.de>
+ *
+ * @copyright Copyright (c) 2016, ownCloud, Inc.
```
Comment (blizzz) [inline apps/user_ldap/lib/LDAPProviderFactory.php]: and here

File: apps/user_ldap/lib/LDAPProvider.php [hunk 5]
```diff
@@ -0,0 +1,187 @@
+<?php
+/**
+ * @author Roger Szabo <roger.szabo@web.de>
+ *
+ * @copyright Copyright (c) 2016, ownCloud, Inc.
```
Comment (blizzz) [inline apps/user_ldap/lib/LDAPProvider.php]: and this

File: apps/user_ldap/lib/LDAPProvider.php [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+	 * Translate a LDAP DN to an ownCloud user name. If there is no mapping between 
+	 * the DN and the user name, a new one will be created.
+	 * @param string $dn LDAP DN
+	 * @return string with the ownCloud user name
+	 * @throws \Exception if translation was unsuccessful
+	 */
+	public function getUserName($dn) {
+		$result = $this->backend->dn2UserName($dn);
+		if(!$result){
+			throw new \Exception('Translation to ownCloud user name unsuccessful');
```
Comment (blizzz) [inline apps/user_ldap/lib/LDAPProvider.php]: replace ``ownCloud`` with ``Nextcloud`` or ``internal``

File: lib/public/LDAP/ILDAPProvider.php [hunk 7]
```diff
@@ -0,0 +1,104 @@
+<?php
+/**
+ * @author Roger Szabo <roger.szabo@web.de>
+ *
+ * @copyright Copyright (c) 2016, ownCloud, Inc.
```
Comment (blizzz) [inline lib/public/LDAP/ILDAPProvider.php]: and this

File: lib/public/LDAP/ILDAPProviderFactory.php [hunk 8]
```diff
@@ -0,0 +1,51 @@
+<?php
+/**
+ * @author Roger Szabo <roger.szabo@web.de>
+ *
+ * @copyright Copyright (c) 2016, ownCloud, Inc.
```
Comment (blizzz) [inline lib/public/LDAP/ILDAPProviderFactory.php]: and this

**Comment by GitHubUser4234** [comment]
thanks a lot @blizzz . Changes are committed accordingly. I have also removed the word "owncloud" from the comments of all **new** API methods.

**Comment by blizzz** [comment]
LGTM

**Comment by blizzz** [comment]
> > > @MorrisJobke: We switched to NC and it is high priority for us to have it included in NC 10.

> > We already released a beta release of NC 10 and will not include new features anymore in the stable10 branch. This needs to wait for NC 11 

> I am not aware we announced a feature freeze, did we? Also, because we still lack other planned features for 10.

Discussed with @karlitschek  earlier this week that a backport would be OK. Since we had no feature freeze, upon to the feature maintainers discretion.

**Comment by blizzz** [comment]
We need a second reviewer :) @MorrisJobke or @schiessle maybe?

**Comment by GitHubUser4234** [comment]
As requested by @schiessle at #600 , new files are updated with nextcloud license headers now.

**Comment by schiessle** [comment]
Can't test it but I'm sure @blizzz and @GitHubUser4234 know what they are doing :wink: 

Had a look at the code, looks good to me :+1:

**Comment by GitHubUser4234** [comment]
Thanks a lot @blizzz @schiessle @MorrisJobke  for making this contribution possible  :+1:

**Pull request merged**